src/interrupts.rs

```diff
@@ -0,0 +1,328 @@
+
+            self.interrupt_type = index;
+            return Ok(());
+        }
+
+        self.interrupt_type = 0;
+        Ok(())
+    }
+}
+
+impl InterruptsQueue {
+    /// Add the `event_fd` file descriptor to epoll.
+    pub fn vfio_epoll_ctl(&mut self, event_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        let mut event: Event = Event {
+            events: libc::EPOLLIN as u32,
+            data: event_fd as u64,
+        };
+
+        let status = unsafe { libc::epoll_create1(0) };
+        if status == -1 {
+            return Err(format!(
+                "failed to epoll_create1. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+        let epoll_fd = status as RawFd;
+
+        if unsafe {
+            libc::epoll_ctl(
+                epoll_fd,
+                libc::EPOLL_CTL_ADD,
+                event_fd,
+                &mut event as *mut _ as *mut libc::epoll_event,
+            )
+        } == -1
+        {
+            return Err(format!(
+                "failed to epoll_ctl. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+
+        self.vfio_epoll_fd = epoll_fd;
+        Ok(())
+    }
+
+    /// Waits for events on the epoll instance referred to by the file descriptor `epoll_fd`.
+    ///
+    /// The memory area pointed to by events will contain the events that will be available for the caller.
+    /// The `timeout` argument specifies the minimum number of milliseconds that epoll_wait will block.
+    /// Specifying a `timeout` of -1 causes epoll_wait to block indefinitely, while specifying a
+    /// `timeout` equal to zero cause epoll_wait to return immediately, even if no events are available.
+    /// Returns the number of ready file descriptors.
+    pub fn vfio_epoll_wait(&self, timeout: i32) -> Result<usize, Box<dyn Error>> {
+        let mut events = [Event::default(); 10];
+        let rc: usize;
+
+        let status = unsafe {
+            libc::epoll_wait(
+                self.vfio_epoll_fd,
+                events.as_mut_ptr() as *mut libc::epoll_event,
+                events.len() as i32,
+                timeout,
+            )
+        };
+        if status == -1 {
+            return Err(format!(
+                "failed to epoll_wait. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+        rc = status as usize;
+        if rc > 0 {
+            /* epoll_wait has at least one fd ready to read */
+            for event in events.iter().take(rc) {
+                let mut val: u16 = 0;
+                let val_ptr: *mut u16 = &mut val;
+                // read event file descriptor to clear interrupt.
+                if unsafe {
+                    libc::read(
+                        event.data as i32,
+                        val_ptr as *mut libc::c_void,
+                        mem::size_of::<u64>(),
```

Comment:
Either this should read `mem::size_of::<u16>()` or `val` should be `u64` ... and `val_ptr` is a bit redundant, just do a `&mut val as *mut _ as *mut libc::c_void` here